### PR TITLE
Implement uv_getrusage_thread

### PR DESCRIPTION
**Pull request by jasnell**
Just a start for an implementation on https://github.com/libuv/libuv/issues/3119

### CODE REVIEW

File: docs/src/misc.rst [hunk 1]
```diff
@@ -262,7 +262,7 @@ API
 .. c:function:: char** uv_setup_args(int argc, char** argv)
 
     Store the program arguments. Required for getting / setting the process title
-    or the executable path. Libuv may take ownership of the memory that `argv` 
+    or the executable path. Libuv may take ownership of the memory that `argv`
```
Comment (santigimeno) [inline docs/src/misc.rst:265]: I'd remove this as it's an unrelated change

File: docs/src/misc.rst [hunk 2]
```diff
@@ -322,6 +322,15 @@ API
...
 
+.. c:function:: int uv_getrusage_thread(uv_rusage_t* rusage)
+
+    Gets the resource usage measures for the calling thread.
+
+    .. note::
+        Not supported on all platforms. May return `UV_ENOTSUP`.
+        On Windows not all fields are set, the unsupported fields are filled with zeroes.
+        See :c:type:`uv_rusage_t` for more details.
+
```
Comment (santigimeno) [inline docs/src/misc.rst:333]: The `.. versionadded::` label should be added

File: docs/src/misc.rst [hunk 3]
```diff
@@ -685,7 +694,7 @@ API
 
     .. note::
         On IBM i PASE, the highest process priority is -10. The constant
-        `UV_PRIORITY_HIGHEST` is -10, `UV_PRIORITY_HIGH` is -7, 
+        `UV_PRIORITY_HIGHEST` is -10, `UV_PRIORITY_HIGH` is -7,
```
Comment (santigimeno) [inline docs/src/misc.rst:697]: Unrelated change

File: src/unix/core.c [hunk 4]
```diff
@@ -998,6 +998,41 @@ int uv_getrusage(uv_rusage_t* rusage) {
   return 0;
 }
 
+int uv_getrusage_thread(uv_rusage_t* rusage) {
+#ifndef RUSAGE_THREAD
```
Comment (santigimeno) [inline src/unix/core.c:1002]: What about having an internal `int uv__getrusage(int who, uv_rusage_t* rusage)` to be called from `uv_getrusage()` and `uv_getrusage_thread()` which implements the actual functionality?

File: src/unix/core.c [hunk 5]
```diff
@@ -998,6 +998,41 @@ int uv_getrusage(uv_rusage_t* rusage) {
...
+  if (getrusage(RUSAGE_THREAD, &usage))
+    return UV__ERR(errno);
+
+  rusage->ru_utime.tv_sec = usage.ru_utime.tv_sec;
+  rusage->ru_utime.tv_usec = usage.ru_utime.tv_usec;
+
+  rusage->ru_stime.tv_sec = usage.ru_stime.tv_sec;
+  rusage->ru_stime.tv_usec = usage.ru_stime.tv_usec;
+
+#if !defined(__MVS__) && !defined(__HAIKU__)
```
Comment (santigimeno) [inline src/unix/core.c:1016]: Do you think we should initialize the `uv_rusage_t` struct to zero `memset(rusage, 0, sizeof(*rusage))` so the fields not supported in some platforms are set to 0?

File: test/test-platform-output.c [hunk 6]
```diff
@@ -88,6 +88,28 @@ TEST_IMPL(platform_output) {
...
+#ifndef RUSAGE_THREAD
+  err = uv_getrusage_thread(&rusage);
+  ASSERT(err = UV_ENOTSUP);
+#else
+  err = uv_getrusage_thread(&rusage);
+  ASSERT(err == 0);
+  ASSERT(rusage.ru_utime.tv_sec >= 0);
+  ASSERT(rusage.ru_utime.tv_usec >= 0);
+  ASSERT(rusage.ru_stime.tv_sec >= 0);
+  ASSERT(rusage.ru_stime.tv_usec >= 0);
```
Comment (santigimeno) [inline test/test-platform-output.c:100]: Could the new assert macros (`ASSERT_EQ`, `ASSERT_UINT64_GE`, etc) be used here?

**Review by santigimeno — CHANGES_REQUESTED** [review]

File: src/unix/core.c [hunk 7]
```diff
@@ -1570,7 +1605,7 @@ int uv__search_path(const char* prog, char* buf, size_t* buflen) {
     buf[*buflen] = '\0';
 
     return 0;
-  } 
+  }
```
Comment (RaisinTen) [inline src/unix/core.c:1608]: Unrelated change

File: src/unix/core.c (hunk 4 above)
Comment (jasnell) [reply src/unix/core.c:1002]: For the non-windows case, that makes sense. There's enough difference with the Windows case that having separate impls is better

File: src/unix/core.c (hunk 5 above)
Comment (jasnell) [reply src/unix/core.c:1016]: The current implementation of uv_getrusage does not do so. I'd rather keep it consistent.

File: docs/src/misc.rst (hunk 2 above)
Comment (jasnell) [reply docs/src/misc.rst]: Added. Hopefully I put the correct value but not sure.

File: src/unix/core.c (hunk 5 above)
Comment (santigimeno) [reply src/unix/core.c]: It seems more like on oversight to me and I think consistency between unices and windows wrt unsupported fields is a good thing to have. Regardless, this can be dealt with in a subsequent PR.

File: docs/src/misc.rst (hunk 2 above)
Comment (santigimeno) [reply docs/src/misc.rst]: It should be `1.42.0` as this is a semver minor change.

File: src/win/util.c [hunk 8]
```diff
@@ -1156,6 +1156,40 @@ int uv_getrusage(uv_rusage_t *uv_rusage) {
   return 0;
 }
 
+int uv_getrusage_thread(uv_rusage_t *uv_rusage) {
+  FILETIME createTime, exitTime, kernelTime, userTime;
+  SYSTEMTIME kernelSystemTime, userSystemTime;
+  int ret;
+
+  ret = GetThreadTimes(GetCurrentThread(), &createTime, &exitTime, &kernelTime, &userTime);
```
Comment (santigimeno) [inline src/win/util.c:1164]: Style: long line (it should be 80 chars max).
While you're here, could you wrap also the same line in `uv_getrusage()`? Thanks!

File: test/test-platform-output.c [hunk 9]
```diff
@@ -88,6 +88,28 @@ TEST_IMPL(platform_output) {
   printf("  maximum resident set size: %llu\n",
          (unsigned long long) rusage.ru_maxrss);
 
+#ifndef RUSAGE_THREAD
+  err = uv_getrusage_thread(&rusage);
+  ASSERT(err = UV_ENOTSUP);
```
Comment (santigimeno) [inline test/test-platform-output.c:93]:
```suggestion
  ASSERT_EQ(err, UV_ENOTSUP);
```

**Review by santigimeno — APPROVED** [review]
LGTM with some nits. CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2130/

**Comment by santigimeno** [comment]
> notably MacOS - although there are other ways to get the per-thread
CPU time on that platform.

@jasnell do you plan adding `MacOS` support at this point?

File: src/unix/core.c [hunk 10]
```diff
@@ -965,11 +965,10 @@ int uv__fd_exists(uv_loop_t* loop, int fd) {
   return (unsigned) fd < loop->nwatchers && loop->watchers[fd] != NULL;
 }
 
-
-int uv_getrusage(uv_rusage_t* rusage) {
+int uv__getrusage(int who, uv_rusage_t* rusage) {
```
Comment (RaisinTen) [inline src/unix/core.c:968]:
```suggestion

int uv__getrusage(int who, uv_rusage_t* rusage) {
```
Style nit: Should we keep a two line gap between these functions like most functions here?

File: src/unix/core.c [hunk 11]
```diff
@@ -998,6 +997,17 @@ int uv_getrusage(uv_rusage_t* rusage) {
   return 0;
 }
 
+int uv_getrusage(uv_rusage_t* rusage) {
```
Comment (RaisinTen) [inline src/unix/core.c:1000]:
```suggestion

int uv_getrusage(uv_rusage_t* rusage) {
```

File: src/unix/core.c [hunk 12]
```diff
@@ -998,6 +997,17 @@ int uv_getrusage(uv_rusage_t* rusage) {
...
+  return uv__getrusage(RUSAGE_SELF, rusage);
+}
+
+int uv_getrusage_thread(uv_rusage_t* rusage) {
```
Comment (RaisinTen) [inline src/unix/core.c:1004]:
```suggestion

int uv_getrusage_thread(uv_rusage_t* rusage) {
```

File: src/unix/core.c [hunk 13]
```diff
@@ -998,6 +997,17 @@ int uv_getrusage(uv_rusage_t* rusage) {
...
+#ifndef RUSAGE_THREAD
+  return UV_ENOTSUP;
+#else
+  return uv__getrusage(RUSAGE_THREAD, rusage);
+#endif
+}
```
Comment (RaisinTen) [inline src/unix/core.c:1010]:
```suggestion
}

```

File: src/win/util.c [hunk 14]
```diff
@@ -1156,6 +1156,40 @@ int uv_getrusage(uv_rusage_t *uv_rusage) {
   return 0;
 }
 
+int uv_getrusage_thread(uv_rusage_t *uv_rusage) {
```
Comment (RaisinTen) [inline src/win/util.c:1159]:
```suggestion

int uv_getrusage_thread(uv_rusage_t *uv_rusage) {
```

File: src/win/util.c [hunk 15]
```diff
@@ -1156,6 +1156,40 @@ int uv_getrusage(uv_rusage_t *uv_rusage) {
...
+                               userSystemTime.wSecond;
+  uv_rusage->ru_utime.tv_usec = userSystemTime.wMilliseconds * 1000;
+
+  uv_rusage->ru_stime.tv_sec = kernelSystemTime.wHour * 3600 +
+                               kernelSystemTime.wMinute * 60 +
+                               kernelSystemTime.wSecond;
+  uv_rusage->ru_stime.tv_usec = kernelSystemTime.wMilliseconds * 1000;
+
+  return 0;
+}
```
Comment (RaisinTen) [inline src/win/util.c:1192]:
```suggestion
}

```

File: src/unix/core.c (hunk 5 above)
Comment (jasnell) [reply src/unix/core.c]: Ok. I also noticed there's no NULL check guard in the existing function. Should there be?

**Comment by jasnell** [comment]
> do you plan adding MacOS support at this point?

Very soon yes. I also want to see if there's a way on windows to provide more of the non-cpu time specific data. Those will come in separate prs, however.

File: docs/src/misc.rst [hunk 16]
```diff
@@ -322,6 +322,17 @@ API
         On Windows not all fields are set, the unsupported fields are filled with zeroes.
         See :c:type:`uv_rusage_t` for more details.
 
+.. c:function:: int uv_getrusage_thread(uv_rusage_t* rusage)
+
+    Gets the resource usage measures for the calling thread.
+
+    .. versionadded:: 1.41.0
```
Comment (vtjnash) [inline docs/src/misc.rst:329]: Now 1.42

**Review by vtjnash — APPROVED** [review]
LGTM, with the style nits already mentioned

**Comment by bnoordhuis** [comment]
I believe this is still blocked on a macos implementation?

**Comment by bnoordhuis** [comment]
This PR has conflicts and has been stale for some time now so I'll close it out but LMK if you want to pick it up again.
